packages/nodes-base/nodes/RabbitMQ/RabbitMQTrigger.node.ts

```diff
@@ -37,136 +38,209 @@ export class RabbitMQTrigger implements INodeType {
-						description: 'Returns only the content property.',
+						description: 'Whether to return only the content property',
 					},
-					...rabbitDefaultOptions,
-				].sort((a, b) => {
-					if ((a as INodeProperties).displayName.toLowerCase() < (b as INodeProperties).displayName.toLowerCase()) { return -1; }
-					if ((a as INodeProperties).displayName.toLowerCase() > (b as INodeProperties).displayName.toLowerCase()) { return 1; }
-					return 0;
-				}) as INodeProperties[],
+				],
 			},
 		],
 	};
 
-
 	async trigger(this: ITriggerFunctions): Promise<ITriggerResponse> {
-		const queue = this.getNodeParameter('queue') as string;
-		const options = this.getNodeParameter('options', {}) as IDataObject;
-
-		const channel = await rabbitmqConnectQueue.call(this, queue, options);
-
-		const self = this;
-
-		const startConsumer = async () => {
-			await channel.consume(queue, async (message: IDataObject) => {
-				if (message !== null) {
-					let content: IDataObject | string = message!.content!.toString();
-
-					const item: INodeExecutionData = {
-						json: {},
-					};
+		const queue = String(this.getNodeParameter('queue'));
+		const queueOptions = fixAssertOptions(this.getNodeParameter('queueOptions', {}));
 
-					if (options.contentIsBinary === true) {
-						item.binary = {
-							data: await this.helpers.prepareBinaryData(message.content),
-						};
+		const options = this.getNodeParameter('options', {});
+		if (!isDataObject(options)) {
+			throw new Error('Unexpected type for options');
+		}
+		const subscriptions = this.getNodeParameter('subscriptions', {});
+		const bindings: IDataObject[] = [];
+		if (isDataObject(subscriptions) && Array.isArray(subscriptions.bindings)) {
+			for (const binding of subscriptions.bindings) {
+				if (isDataObject(binding)) {
+					bindings.push(binding);
+				}
+			}
+		}
 
-						item.json = message;
-						message.content = undefined;
-					} else {
-						if (options.jsonParseBody === true) {
-							content = JSON.parse(content as string);
+		const channel = await rabbitmqConnect(this, async (channel) => {
+			await channel.assertQueue(queue, queueOptions);
+			await channel.prefetch(Number(options?.prefetch || 100));
+			for (const { exchange, pattern } of bindings) {
+				await channel.bindQueue(queue, String(exchange), String(pattern));
+			}
+			// we can't unbind other exchanges because amqp doesn't expose a listing
+		});
+
+		const handler = async (message: ConsumeMessage) => {
+			if (message !== null) {
+				messageToItem(this, message, options)
+					.then((item) => {
+						this.emit([[item]]);
+						channel.ack(message);
+					})
+					.catch((error) => {
+						console.error(`Error consuming RabbitMQ message: ${error.message || error}`);
+						// TODO should we nack it? It probably won't succeed on a reattempt, but it wasn't successful either
+						// FIXME how do we report a failure to the ux?
+						if(options?.nack) {
+							channel.nack(message, false, false);
```

Comment:
I'm not sure if this should be used. Normally you'd `nack` to put it back in the queue so it's reattempted, but would there be a case where that would be helpful? If it errors out on the message parsing, it's not going to succeed on a second attempt right?

If we could nack the message if the workflow failed somewhere down the line, that'd probably be ideal, but I couldn't see a way to do that.

Reply:
Oh, nice. I didn't catch that. In that case, I don't see a reason that you wouldn't want to `nack` it, so you probably don't need the option. It should just always send that nack if it can't parse a message.

Does that deferred promise on the `emit` normally get the response of the last node in the workflow, and reject it if the workflow fails anywhere? I can take a look at implementing something like that for the emit later, but it's probably not needed to merge this in.

Reply:
> I'm not sure if this should be used. Normally you'd `nack` to put it back in the queue so it's reattempted, but would there be a case where that would be helpful? If it errors out on the message parsing, it's not going to succeed on a second attempt right?
> 
> If we could nack the message if the workflow failed somewhere down the line, that'd probably be ideal, but I couldn't see a way to do that.

Just to be sure I understand: the aim would be to be able to wait until the end of the workflow, like for the webhook trigger

![image](https://user-images.githubusercontent.com/128426/189313388-78810340-5b72-471f-a815-6e62eb1e0d5f.png)

and based on the workflow's result (or some data returned by the last node in the workflow), decide to ack or nack the message, right?

if I understood it right, yes @Ken-Michalak that would be awesome if you could take a look and super useful for our use case